File: dart_payway/lib/src/models/payway_merchant.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'dart:convert';

class PaywayMerchant {
  /// [internal]
  /// provided by api
  final String merchantID;
  final String merchantApiKey;
  final String merchantApiName;
  final String baseApiUrl;
  final String refererDomain;

  /// ## [PaywayMerchant]
  /// `Represent and Hold Merchant Credential provided by aba bank supporter`
  /// ### [Example]
  /// ```
  ///var merchant = PaywayMerchant(
  ///   merchantID: "your_merchant_api",
  ///   merchantApiKey: "your_api_key",
  ///   merchantApiName: "your_merchant_name",
  ///   baseApiUrl: "based_api_url", // without merchantApiName
  ///   referedDomain: "", // whitelist domain
  /// );
  /// ```
  ///
  PaywayMerchant({
    required this.merchantID,
    required this.merchantApiKey,
    required this.merchantApiName,
    required this.baseApiUrl,
    required this.refererDomain,
  });

  PaywayMerchant copyWith({
    String? merchantID,
    String? merchantApiKey,
    String? merchantApiName,
    String? baseApiUrl,
    String? refererDomain,
  }) {
    return PaywayMerchant(
      merchantID: merchantID ?? this.merchantID,
      merchantApiKey: merchantApiKey ?? this.merchantApiKey,
      merchantApiName: merchantApiName ?? this.merchantApiName,
      baseApiUrl: baseApiUrl ?? this.baseApiUrl,
      refererDomain: refererDomain ?? this.refererDomain,
    );
  }

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'merchantID': merchantID,
      'merchantApiKey': merchantApiKey,
      'merchantApiName': merchantApiName,
      'baseApiUrl': baseApiUrl,
      'refererDomain': refererDomain,
    };
  }

  factory PaywayMerchant.fromMap(Map<String, dynamic> map) {
    return PaywayMerchant(
      merchantID: map['merchantID'] as String,
      merchantApiKey: map['merchantApiKey'] as String,
      merchantApiName: map['merchantApiName'] as String,
      baseApiUrl: map['baseApiUrl'] as String,
      refererDomain: map['refererDomain'] as String,
    );
  }

  String toJson() => json.encode(toMap());

  factory PaywayMerchant.fromJson(String source) =>
      PaywayMerchant.fromMap(json.decode(source) as Map<String, dynamic>);

  @override
  String toString() {
    return 'PaywayMerchant(merchantID: $merchantID, merchantApiKey: $merchantApiKey, merchantApiName: $merchantApiName, baseApiUrl: $baseApiUrl, refererDomain: $refererDomain)';
  }

  @override
  bool operator ==(covariant PaywayMerchant other) {
    if (identical(this, other)) return true;

    return other.merchantID == merchantID &&
        other.merchantApiKey == merchantApiKey &&
        other.merchantApiName == merchantApiName &&
        other.baseApiUrl == baseApiUrl &&
        other.refererDomain == refererDomain;
  }

  @override
  int get hashCode {
    return merchantID.hashCode ^
        merchantApiKey.hashCode ^
        merchantApiName.hashCode ^
        baseApiUrl.hashCode ^
        refererDomain.hashCode;
  }
}

File: php_payway/src/models/PaywayMerchant.php
<?php

namespace PhpPayway\Models;

class PaywayMerchant
{
    public function __construct(
        public string $merchantID,
        public string $merchantApiKey,
        public string $merchantApiName,
        public string $baseApiUrl,
        public string $refererDomain,
    )
    {

    }

    public function copyWith(
        ?string $merchantID = null,
        ?string $merchantApiKey = null,
        ?string $merchantApiName = null,
        ?string $baseApiUrl = null,
        ?string $refererDomain = null,
    )
    {
        $copy = clone($this);

        $this->merchantID = $merchantID ?? $copy->merchantID;
        $this->merchantApiKey = $merchantApiKey ?? $copy->merchantApiKey;
        $this->merchantApiName = $merchantApiName ?? $copy->merchantApiName;
        $this->baseApiUrl = $baseApiUrl ?? $copy->baseApiUrl;
        $this->refererDomain = $refererDomain ?? $copy->refererDomain;
    }

    public function toMap(): array
    {
        return [
            'merchantID' => $this->merchantID,
            'merchantApiKey' => $this->merchantApiKey,
            'merchantApiName' => $this->merchantApiName,
            'baseApiUrl' => $this->baseApiUrl,
            'refererDomain' => $this->refererDomain,
        ];
    }
}